File: leetcode/blind-75/cpp-solutions/longest_substring_no_duplicates.cpp
#include <unordered_map>
#include <string>
#include <iostream>
using namespace std;

//Given a string, find the length of the longest substring with no duplicate characters.
class Solution {
public:
  int lengthOfLongestSubstring(string s){
    unordered_map<char, int> lastIndex; // Maps each char to its most recent index
    int maxLen = 0;
    int start = 0;

    // Expand the sliding window[start, end]
    for(int end = 0; end < s.length(); ++end) {
        char c = s[end];

        // If 'c' was seen before and is within the current window, move start right after its prev occurrence.
        if (lastIndex.count(c) && lastIndex[c] >= start) {
            start = lastIndex[c] + 1;
        }
      
        // Update the last seen index of 'c'.
        lastIndex[c] = end;
      
        // Update the max len if current window is longer.
        maxLen = max(maxLen, end - start + 1);
    }
    return maxLen;
  }

File: leetcode/grind-75/python-solutions/longest_substring_no_duplicates.py
class Solution:
  def lengthOfLongestSubstring(self, s:str) -> int:
    last_index = {}
    max_len = 0
    start = 0
    for end, c in enumerate(s):
      if c in last_index and last_index[c] >= start:
        start = last_index[c]+1
      last_index[c] = end
      max_len = max(max_len, end - start + 1)
    return max_len
